Mathlib/Order/Birkhoff.lean

```diff
@@ -0,0 +1,228 @@
+/-
+Copyright (c) 2022 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import Mathlib.Data.Fintype.Order
+import Mathlib.Order.Irreducible
+import Mathlib.Data.Finset.LocallyFinite
+import Mathlib.Order.UpperLower.Basic
+
+/-!
+# Birkhoff's representation theorem
+
+This file proves the Birkhoff representation theorem: Any finite distributive lattice can be
+represented as a sublattice of some powerset algebra.
+
+Precisely, any nonempty finite distributive lattice is isomorphic to the lattice of lower sets of
+its irreducible elements. And conversely it is isomorphic to the order of its irreducible lower
+sets.
+
+## Main declarations
+
+For a nonempty finite distributive lattice `α`:
+* `OrderIso.lowerSetSupIrred`: `α` is isomorphic to the lattice of lower sets of its irreducible
+  elements.
+* `OrderIso.supIrredLowerSet`: `α` is isomorphic to the order of its irreducible lower sets.
+* `OrderEmbedding.birkhoffSet`, `OrderEmbedding.birkhoffFinet`: Order embedding of `α` into the
+  powerset lattice of its irreducible elements.
+* `LatticeHom.birkhoffSet`, `LatticeHom.birkhoffFinet`: Same as the previous two, but bundled as
+  an injective lattice homomorphism.
+* `exists_birkhoff_representation`: `α` embeds into some powerset algebra. You should prefer using
+  this over the explicit Birkhoff embedding because the Birkhoff embedding is littered with
+  decidability footguns that this existential-packaged version can afford to avoid.
+
+## See also
+
+This correspondance between finite distributive lattices and finite boolean algebras is made
+functorial in... TODO: Actually do it.
+
+## Tags
+
+birkhoff, representation, stone duality, lattice embedding
+-/
+
+open Finset Function OrderDual
+
+variable {α : Type*}
+
+namespace UpperSet
+variable [SemilatticeInf α] {s : UpperSet α} {a : α}
+
+@[simp] lemma infIrred_Ici (a : α) : InfIrred (Ici a) := by
+  refine' ⟨fun h ↦ Ici_ne_top h.eq_top, fun s t hst ↦ _⟩
```

Comment:
I had the impression `refine'` is somewhat deprecated, and we're supposed to use `refine` and explicitly indicate the holes we want to make goals.
```suggestion
  refine ⟨fun h ↦ Ici_ne_top h.eq_top, fun s t hst ↦ ?_⟩
```